File: app/src/main/java/orinasa/njarasoa/maripanatokana/data/remote/WmoWeatherCode.kt
package orinasa.njarasoa.maripanatokana.data.remote

import androidx.annotation.StringRes
import orinasa.njarasoa.maripanatokana.R

/**
 * Maps WMO weather interpretation codes to string resource IDs.
 * https://open-meteo.com/en/docs#weathervariables
 */
@StringRes
fun wmoDescriptionRes(code: Int): Int = when (code) {
    0 -> R.string.wmo_clear_sky
    1 -> R.string.wmo_mainly_clear
    2 -> R.string.wmo_partly_cloudy
    3 -> R.string.wmo_overcast
    45 -> R.string.wmo_fog
    48 -> R.string.wmo_rime_fog
    51 -> R.string.wmo_light_drizzle
    53 -> R.string.wmo_moderate_drizzle
    55 -> R.string.wmo_dense_drizzle
    56 -> R.string.wmo_light_freezing_drizzle
    57 -> R.string.wmo_dense_freezing_drizzle
    61 -> R.string.wmo_slight_rain
    63 -> R.string.wmo_moderate_rain
    65 -> R.string.wmo_heavy_rain
    66 -> R.string.wmo_light_freezing_rain
    67 -> R.string.wmo_heavy_freezing_rain
    71 -> R.string.wmo_slight_snowfall
    73 -> R.string.wmo_moderate_snowfall
    75 -> R.string.wmo_heavy_snowfall
    77 -> R.string.wmo_snow_grains
    80 -> R.string.wmo_slight_rain_showers
    81 -> R.string.wmo_moderate_rain_showers
    82 -> R.string.wmo_violent_rain_showers
    85 -> R.string.wmo_slight_snow_showers
    86 -> R.string.wmo_heavy_snow_showers
    95 -> R.string.wmo_thunderstorm
    96 -> R.string.wmo_thunderstorm_slight_hail
    99 -> R.string.wmo_thunderstorm_heavy_hail
    else -> R.string.wmo_unknown
}

fun wmoEmoji(code: Int, isNight: Boolean = false): String = when (code) {
    0 -> if (isNight) "\uD83C\uDF11"  // 🌑
         else "\u2600\uFE0F"           // ☀️
    1 -> if (isNight) "\uD83C\uDF14"  // 🌔
         else "\uD83C\uDF24\uFE0F"    // 🌤️
    2 -> if (isNight) "\uD83C\uDF13"  // 🌓
         else "\u26C5"                 // ⛅
    3 -> "\u2601\uFE0F"       // ☁️
    45, 48 -> "\uD83C\uDF2B\uFE0F" // 🌫️
    51, 53, 55 -> "\uD83C\uDF26\uFE0F" // 🌦️
    56, 57 -> "\uD83C\uDF28\uFE0F" // 🌨️
    61, 63 -> "\uD83C\uDF27\uFE0F" // 🌧️
    65 -> "\uD83C\uDF27\uFE0F" // 🌧️
    66, 67 -> "\uD83C\uDF28\uFE0F" // 🌨️
    71, 73, 75, 77 -> "\u2744\uFE0F" // ❄️
    80, 81, 82 -> "\uD83C\uDF26\uFE0F" // 🌦️
    85, 86 -> "\uD83C\uDF28\uFE0F" // 🌨️
    95 -> "\u26C8\uFE0F"      // ⛈️
    96, 99 -> "\u26C8\uFE0F"  // ⛈️
    else -> "\uD83C\uDF10"    // 🌐
}

/**
 * Maps WMO weather code + is_day to an OWM-style icon code for backward compatibility.
 */
fun wmoIconCode(code: Int, isDay: Boolean): String {
    val base = when (code) {
        0 -> "01"
        1 -> "02"
        2 -> "03"
        3 -> "04"
        45, 48 -> "50"
        51, 53, 55, 56, 57 -> "09"
        61, 63, 65, 66, 67 -> "10"
        71, 73, 75, 77, 85, 86 -> "13"
        80, 81, 82 -> "09"
        95, 96, 99 -> "11"
        else -> "01"
    }
    val suffix = if (isDay) "d" else "n"
    return "$base$suffix"
}

File: react-native/src/data/WmoWeatherCode.ts
/** Maps WMO weather interpretation codes to i18n string keys. */
export function wmoDescriptionKey(code: number): string {
  switch (code) {
    case 0: return 'wmo_clear_sky';
    case 1: return 'wmo_mainly_clear';
    case 2: return 'wmo_partly_cloudy';
    case 3: return 'wmo_overcast';
    case 45: return 'wmo_fog';
    case 48: return 'wmo_rime_fog';
    case 51: return 'wmo_light_drizzle';
    case 53: return 'wmo_moderate_drizzle';
    case 55: return 'wmo_dense_drizzle';
    case 56: return 'wmo_light_freezing_drizzle';
    case 57: return 'wmo_dense_freezing_drizzle';
    case 61: return 'wmo_slight_rain';
    case 63: return 'wmo_moderate_rain';
    case 65: return 'wmo_heavy_rain';
    case 66: return 'wmo_light_freezing_rain';
    case 67: return 'wmo_heavy_freezing_rain';
    case 71: return 'wmo_slight_snowfall';
    case 73: return 'wmo_moderate_snowfall';
    case 75: return 'wmo_heavy_snowfall';
    case 77: return 'wmo_snow_grains';
    case 80: return 'wmo_slight_rain_showers';
    case 81: return 'wmo_moderate_rain_showers';
    case 82: return 'wmo_violent_rain_showers';
    case 85: return 'wmo_slight_snow_showers';
    case 86: return 'wmo_heavy_snow_showers';
    case 95: return 'wmo_thunderstorm';
    case 96: return 'wmo_thunderstorm_slight_hail';
    case 99: return 'wmo_thunderstorm_heavy_hail';
    default: return 'wmo_unknown';
  }
}

/** Returns emoji based on WMO weather code + time of day. */
export function wmoEmoji(code: number, isNight: boolean = false): string {
  switch (code) {
    case 0:
      return isNight ? '\uD83C\uDF11' : '\u2600\uFE0F';     // 🌑 or ☀️
    case 1:
      return isNight ? '\uD83C\uDF14' : '\uD83C\uDF24\uFE0F'; // 🌔 or 🌤️
    case 2:
      return isNight ? '\uD83C\uDF13' : '\u26C5';            // 🌓 or ⛅
    case 3:
      return '\u2601\uFE0F';          // ☁️
    case 45:
    case 48:
      return '\uD83C\uDF2B\uFE0F';    // 🌫️
    case 51:
    case 53:
    case 55:
      return '\uD83C\uDF26\uFE0F';    // 🌦️
    case 56:
    case 57:
      return '\uD83C\uDF28\uFE0F';    // 🌨️
    case 61:
    case 63:
    case 65:
      return '\uD83C\uDF27\uFE0F';    // 🌧️
    case 66:
    case 67:
      return '\uD83C\uDF28\uFE0F';    // 🌨️
    case 71:
    case 73:
    case 75:
    case 77:
      return '\u2744\uFE0F';          // ❄️
    case 80:
    case 81:
    case 82:
      return '\uD83C\uDF26\uFE0F';    // 🌦️
    case 85:
    case 86:
      return '\uD83C\uDF28\uFE0F';    // 🌨️
    case 95:
    case 96:
    case 99:
      return '\u26C8\uFE0F';          // ⛈️
    default:
      return '\uD83C\uDF10';          // 🌐
  }
}

/** Maps WMO weather code + is_day to an OWM-style icon code. */
export function wmoIconCode(code: number, isDay: boolean): string {
  let base: string;
  switch (code) {
    case 0: base = '01'; break;
    case 1: base = '02'; break;
    case 2: base = '03'; break;
    case 3: base = '04'; break;
    case 45: case 48: base = '50'; break;
    case 51: case 53: case 55: case 56: case 57: base = '09'; break;
    case 61: case 63: case 65: case 66: case 67: base = '10'; break;
    case 71: case 73: case 75: case 77: case 85: case 86: base = '13'; break;
    case 80: case 81: case 82: base = '09'; break;
    case 95: case 96: case 99: base = '11'; break;
    default: base = '01'; break;
  }
  return `${base}${isDay ? 'd' : 'n'}`;
}
